modules/xfeatures2d/src/freak.cpp

```diff
@@ -333,8 +333,12 @@ void FREAK_Impl::compute( InputArray _image, std::vector<KeyPoint>& keypoints, O
     ((FREAK_Impl*)this)->buildPattern();
 
     // Convert to gray if not already
-    Mat grayImage = image;
-    CV_Assert(grayImage.channels() == 1);
+    Mat grayImage;
+    if( image.channels() == 3 || image.channels() == 4 )
+        cvtColor(image, grayImage, COLOR_BGR2GRAY);
+    else
+        CV_Assert(image.channels() == 1);
+        grayImage = image;
```

Comment:
Assignment is now out of else branch scope so in case image.channels() == 3 it will overwrite produced greyscale image back to RGB

Reply:
Excuse my Python habits! Thank you..